[ML] Add prefix strings option to trained models

davidkyle opened the pull request:
Certain NLP models such as [multilingual-e5-large](https://huggingface.co/intfloat/multilingual-e5-large) require a prefix string to be applied to the input text. For asymmetric tasks such as information retrieval the prefix can be different when ingesting the data and when searching it. For example text embedding model can have a one prefix applied when the model is evaluated as part of an knn search and a different prefix when ingesting documents. 

An example model configuration with prefix strings is:
```
{
    "model_type": "pytorch",
    "inference_config": {
        "text_embedding": {
            "tokenization": {
                "xlm_roberta": {
                    "do_lower_case": false,
                    "with_special_tokens": true,
                    "max_sequence_length": 512,
                    "truncate": "first",
                    "span": -1
                }
            },
            "embedding_size": 384
        }
    },
    "prefix_strings": {
        "search": "this is a query",
        "ingest": "this is a passage"
    }
}
```

Many files have been touched by this change but the bulk of the work is quite simple: define the configuration object an pass the prefix type (context) parameter down the inference calls.

davidkyle committed:
Prefix strings

davidkyle committed:
Prefix strings

davidkyle committed:
fix the tests

davidkyle committed:
prefix type in the request

DaveCTurner committed:
Avoid negative DesiredBalanceStats#lastConvergedIndex (#101998)

The initial state of the desired-balance allocator has
`lastConvergedIndex` set to `-1`. This is not important to represent in
the stats, so with this commit we map it to zero.

abdonpijpelink committed:
[DOCS] DISSECT does not support reference keys (#102002)

jimczi committed:
Corrected the expansion of overlapping terms in the unified highlighter (#101912)

This commit addresses an issue in the passage formatter of the unified highlighter, where overlapping terms were not correctly expanded to be highlighted as a single object. The fix in this commit involves adjusting the expansion logic to consider the maximum end offset during the process, as matches are initially sorted by ascending start offset and then by ascending end offset.

original-brownbear committed:
Mute tests for #102000 (#102006)

It's in the title these two tests are failing near 100% of the time.

for #102000

davidkyle committed:
docs

elasticsearchmachine commented:
Hi @davidkyle, I've created a changelog YAML for you.

davidkyle committed:
Update docs/changelog/101978.yaml

szabosteve commented on docs/reference/ml/trained-models/apis/put-trained-models.asciidoc:
```diff
@@ -701,6 +701,32 @@ the platform identifiers used by Elasticsearch, so one of, `linux-x86_64`,
+`search`:::
+(Optional, string)
+The prefix string to prepend to the input text for requests
+originating from a search query.
```
```suggestion
originating from a search query.

```

szabosteve commented on docs/reference/ml/trained-models/apis/put-trained-models.asciidoc:
```diff
@@ -701,6 +701,32 @@ the platform identifiers used by Elasticsearch, so one of, `linux-x86_64`,
+`ingest`:::
+(Optional, string)
+The prefix string to prepend to the input text for requests
+at ingest where the Inference ingest processor is used. // TODO is there a shortcut for Inference ingest processor?
```
```suggestion
at ingest where the {infer} ingest processor is used. // TODO is there a shortcut for Inference ingest processor?
```

szabosteve replied:
There is no attribute to shorten `inference ingest processor`. Do you want me to create one? Or `{infer}` for inference is enough?

original-brownbear committed:
Mute test for #102010 (#102011)

Muting this one that keeps failing for #102010

piergm committed:
Remove some more explicit SearchResponse use in tests (#102003)

The title says the whole story

droberts195 reviewed (approved):
LGTM, but if we're going to have packaged models on GCS that use this feature then the new section will need to also go in this class: https://github.com/elastic/elasticsearch/blob/8d6ded3736ec12dc56d2edd6f0d493916debbc45/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/ModelPackageConfig.java#L33

masseyke committed:
Allowing non-dynamic index settings to be updated by automatically unassigning shards (#101723)

gmarouli committed:
ES-6566: Move the calculation of data tier usage stats to individual nodes (#100230) (#101599)

brianseeders committed:
[ci] Fix build scan annotations on Windows (#101990)

benwtrent committed:
Remove all SearchResponse direct references from tests (#102024)

Moving towards ref counted search responses, this removes all direct
references to SearchResponse from ML tests.

The remaining references in ML tests are for mocked search responses.

Additionally, while making this change, I noticed there are multiple
places in ML production code that have the `SearchResponse
searchResponse = ` pattern. Meaning, once we refCount `SearchResponse`,
those places will have to be updated as well (in a future PR).

Related to: https://github.com/elastic/elasticsearch/pull/100966 Related
to: https://github.com/elastic/elasticsearch/issues/99590

breskeby committed:
Revert "Keep old plugin name for build complete plugin (#101995)" (#102026)

This reverts commit 46f545cf458c4a304046d2bced66f61f878f683e as we need
to sort out the cause of #102022 first

original-brownbear committed:
Remove explicit SearchResponse references from server metrics aggs ITs (#101999)

Remove explicit references from
`org.elasticsearch.search.aggregations.metrics`

closes #101766 

Co-authored-by: Ignacio Vera <ivera@apache.org>

original-brownbear committed:
Fix a few more explicit SearchResponse uses in the server package (#102031)

For #102030

gmarouli committed:
Revert "ES-6566: Move the calculation of data tier usage stats to individual nodes (#100230) (#101599)" (#102042)

Reverting because the new action is not properly handled in a mixed
cluster.

breskeby committed:
Fix caching for StandaloneRestIntegTest tasks (#102043)

This fixes running caching issues for StandaloneRestIntegTest tasks as
an aftermath of https://github.com/elastic/elasticsearch/pull/101923

This should fix #102015

mark-vieira committed:
Mute UpgradeClusterClientYamlTestSuiteIT

mark-vieira committed:
Revert "Mute UpgradeClusterClientYamlTestSuiteIT"

This reverts commit e9e948d9d848d66d57f57cd5114f3a526fec3c2a.

original-brownbear committed:
Remove explicit search responses from IndexAliasIT (#101996)

Just like the other ones.

masseyke committed:
Removing the use of Version.CURRENT from watcher (#102045)

piergm committed:
null check on searchResponse (#102017)

closes #102016

JVerwolf committed:
Update tests to decrement ref count  (#102044)

Part of the broader work covered in
https://github.com/elastic/elasticsearch/issues/102030

Updates tests in: - HighlighterWithAnalyzersTests.java -
TokenCountFieldMapperIntegrationIT - GeoIpDownloaderIT.java -
DataStreamIT.java

iverase committed:
Remove explicit SearchResponse references from server bucket aggs (part1) (#102035)

iverase committed:
Remove explicit SearchResponse references from server bucket aggs (part2) (#102036)

iverase committed:
Remove explicit SearchResponse references from server bucket aggs (part3) (#102037)

iverase committed:
Remove explicit SearchResponse references from server bucket aggs (part4) (#102038)

DaveCTurner committed:
Repo analysis: verify empty register (#102048)

We encountered a bad third-party S3 repository implementation which
incorrectly rejects empty multipart uploads. This anomaly is detected by
some repository analysis runs, but not by all of them. This commit adds
a specific check for this incompatibility so that it can be reported
reliably.

DaveCTurner committed:
Repo analysis: allow configuration of register ops (#102051)

Adds the `?register_operation_count` parameter that allows to control
the number of register operations separately from the number of regular
blob operations.

jan-elastic committed:
Add stats by model to the machine learning usage stats. (#101915)

* Add inference counts by NLP model to the machine learning usage stats.

* Update docs/changelog/101915.yaml

* Add inference_counts_by_model to yamlRestTest.

* Strip leading dot from internal model IDs.

* Add last access and task type to the stats by model.

* Change stats_by_model for map to list

* Simplify code.

* Fix style

ldematte committed:
Using VersionInformation explicitly (#101962)

piergm committed:
Remove some more explicit SearchResponse use in tests Pt2 (#102029)

The title says the whole story part 2

abdonpijpelink committed:
[DOCS] Clarify ES|QL grok escaping (#102059)

andreidan committed:
[DOCS] Fix typo (#101791)

thecoop committed:
Remove obsolete version check for autoid creates (#101623)

thecoop committed:
Update connector templates to use a historical feature (#101531)

pmpailis committed:
Removing explicit SearchResponse usages in tests - v3 (#102019)

Tests covered in this PR:

* `org.elasticsearch.percolator.PercolatorQuerySearchIT`

davidkyle committed:
Update docs/reference/ml/trained-models/apis/put-trained-models.asciidoc

Co-authored-by: István Zoltán Szabó <istvan.szabo@elastic.co>

davidkyle committed:
Update docs/reference/ml/trained-models/apis/put-trained-models.asciidoc

Co-authored-by: István Zoltán Szabó <istvan.szabo@elastic.co>

szabosteve reviewed (approved):
Docs LGTM, thanks!

davidkyle commented:
@elasticmachine update branch

elasticmachine commented:
merge conflict between base and head

DaveCTurner committed:
AwaitsFix for #102070

pmpailis committed:
Removing explicit SearchResponse usages in tests (#102008)

Tests covered in this PR: *
`org.elasticsearch.xpack.enrich.EnrichPolicyRunnerTests` *
`org.elasticsearch.index.engine.frozen.FrozenIndexIT` *
`org.elasticsearch.index.engine.frozen.FrozenIndexTests`

davidkyle committed:
Add prefix strings to model package config

davidkyle commented:
Closing as I've made a mess with git. #102089 is raised as a replacement
